refactor(default-flatpaks): Switch to standardized BlueBuild location, implement useful docs into files & make flatpak detection + comparison more robust

fiftydinar opened the pull request:
Use

`/usr/etc/bluebuild/default-flatpaks`

location instead of

`/usr/etc/flatpaks`

If possible, we should ideally use this location for system modifications:

`/usr/share/bluebuild/default-flatpaks`

While having user modifications in:

`/usr/etc/bluebuild/default-flatpaks`

But it needs to be figured out how the logic will work for separating system & user modifications this way.

I used this logic in unofficial `initramfs-setup` module & it works well, by merging both system & user files into 1 output.

However, this method can create duplicates if user specified it in it's modification, so I mentioned that user should look if the system entry has modifications they need. Perhaps, array diff can be done, which would circumvent this.

fiftydinar committed:
chore(default-flatpaks): Switch to standardized BlueBuild location

Use

`/usr/etc/bluebuild/default-flatpaks`

location instead of

`/usr/etc/flatpaks`

If possible, we should ideally use this location for system modifications:

`/usr/share/bluebuild/default-flatpaks`

While having user modifications in:

`/usr/etc/bluebuild/default-flatpaks`

But it needs to be figured out how the logic will work for separating system & user modifications this way.

I used this logic in unofficial `initramfs-setup` module & it works well, by merging both system & user files into 1 output.

However, this method can create duplicates if user specified it in it's modification, so I mentioned that user should look if the system entry has modifications they need. Perhaps, array diff can be done, which would circumvent this.

So merge this for start.

Only other module which has potential for migrating to standardized BlueBuild config is yafti (more details in other PR I'll do in some time).

xynydev commented:
> Only other module which has potential for migrating to standardized BlueBuild config is yafti (more details in other PR I'll do in some time).

I'm not sure about that, as `yafti` is very clearly _Universal Blue_. I'll follow up with this more later.

fiftydinar committed:
refactor(default-flatpaks): Make default-flatpaks more robust

Document in detail to user how he can change the config in files itself.

Also document what files do in `/usr/share/bluebuild/default-flatpaks` as well.

Refactor flatpak lists detection to be more reliable by excluding words starting with # symbols, whitelines & duplicate entries. Use `comm` for comparing flatpak list to existing flatpaks output instead of using `grep`, as it's easier to use & it's more reliable.

Separate user's & maintainer's modifications better by utilizing read-only `/usr/share/bluebuild/default-flatpaks` directory for maintainers, while for users we would use `(/usr)/etc/bluebuild/default-flatpaks` directory. Reverting to defaults is more reliable as it would avoid users from touching maintainer's modification directly.

I wouldn't modify repo-info.yml doc content, as we restrict it from user's modification & we wouldn't want to potentially ruin yaml parsing just for that.

Only thing that remains is to test this in a VM. And look to potentially make code cleaner.

fiftydinar commented:
> > Only other module which has potential for migrating to standardized BlueBuild config is yafti (more details in other PR I'll do in some time).
> 
> I'm not sure about that, as `yafti` is very clearly _Universal Blue_. I'll follow up with this more later.

No need for me, it's clear. I thought we could migrate yafti.yml at least, but I decided that it's the best to leave that as it is.

fiftydinar commented:
I refactored code now to achieve what I wanted here, but I also added useful docs into files used by `default-flatpaks` & I made flatpak lists detection & comparison to current flatpaks more reliable now.

Here's the commit description:

Document in detail to user how he can change the config in files itself.

Also document what files do in `/usr/share/bluebuild/default-flatpaks` as well.

Refactor flatpak lists detection to be more reliable by excluding words starting with # symbols, whitelines & duplicate entries. Use `comm` for comparing flatpak list to existing flatpaks output instead of using `grep`, as it's easier to use & it's more reliable.

Separate user's & maintainer's modifications better by utilizing read-only `/usr/share/bluebuild/default-flatpaks` directory for maintainers, while for users we would use `(/usr)/etc/bluebuild/default-flatpaks` directory. Reverting to defaults is more reliable as it would avoid users from touching maintainer's modification directly.

I wouldn't modify repo-info.yml doc content, as we restrict it from user's modification & we wouldn't want to potentially ruin yaml parsing just for that.

Only thing that remains is to test this in a VM. And look to potentially make code cleaner.

fiftydinar committed:
draft(default-flatpaks): Avoid install/remove duplicate loop scenario

I have to figure out this part

fiftydinar committed:
draft(default-flatpaks): Avoid install/remove duplicate loop scenario pt.2

fiftydinar committed:
draft(default-flatpaks): Avoid install/remove duplicate loop scenario pt.3

fiftydinar committed:
fix(default-flatpaks): Avoid install/remove duplicate loop scenario

fiftydinar committed:
chore(default-flatpaks): minor grammar adjustment

fiftydinar committed:
draft(default-flatpaks): Account for scenario...

when user sets the same package in install list that is in maintainer's remove list & vice-versa

fiftydinar committed:
fix(default-flatpaks): Account for scenario when user sets the same package in install list that is in maintainer's remove list & vice-versa

fiftydinar committed:
fix(default-flatpaks): Typo in code for combined install list

fiftydinar committed:
chore(default-flatpaks): Remove unnecessary echo

fiftydinar committed:
chore(default-flatpaks): Make directory for user config

fiftydinar committed:
chore(default-flatpaks): Explain user's vs maintainer's flatpak list situation better

fiftydinar commented:
I made custom image for this PR based on this latest commit:

`rpm-ostree rebase ostree-unverified-registry:ghcr.io/fiftydinar/default-flatpaks-refactor:39`

So it is easier for others if they want to test this too.

It's based on Silverblue-main, installs those system flatpaks:
- org.gnome.Boxes
- org.gnome.Calculator

Removes those system flatpaks:
- org.gnome.Loupe
- org.gnome.Calendar

That's it.

fiftydinar committed:
fix(default-flatpaks): Typo in user's configuration for system flatpaks install

fiftydinar commented:
I tested these scenarios & it works well:

- classic maintainer's install & remove list
- user's remove list on top of maintainer's
- user's install list on top of maintainer's
- user's install + remove list on top of maintainer's

#

- classic maintainer's notifications configuration
- user's notifications configuration

#

- duplicate entries detection

#

I find this refactor done, so I will mark this PR as ready:

fiftydinar committed:
chore(default-flatpaks): There is no need to mkdir parent folder if child folder is created

xynydev commented:
Looks great! Haven't got time for an in-depth review right now, but I promise one soon. More general comments on this topic at https://github.com/blue-build/modules/issues/157

xynydev commented on modules/default-flatpaks/default-flatpaks.sh:
```diff
@@ -103,23 +103,98 @@ configure_lists () {
     configure_flatpak_repo "$1" "system"
+    system_install_list_doc="/usr/share/bluebuild/default-flatpaks/system/install"
+    system_remove_list_doc="/usr/share/bluebuild/default-flatpaks/system/remove"
+    echo -e "# This file utilizes maintainer's configuration for \`system flatpaks install\` used by \`default-flatpaks\` BlueBuild module.
```
These docs echoes should be moved to the correct files with the correct file structure directly in the module source. Just create a new directory called like `config` or smn with the correct file structure and docs inside files, then just copy it into place in the script.

fiftydinar replied:
> These docs echoes should be moved to the correct files with the correct file structure directly in the module source. Just create a new directory called like `config` or smn with the correct file structure and docs inside files, then just copy it into place in the script.

I thought about if I should change those to files or not, but I waited for your review.

xynydev replied:
Would you do that before I make my changes?

fiftydinar replied:
@xynydev Done

xynydev reviewed (changes requested):
Basically the rest of my change reqs in this PR would be documentation phrasing and grammar, which I find would be too laborious to spell out here individually. I'll do a commit for the PR fixing those issues later, if that's alright.

fiftydinar commented:
> Basically the rest of my change reqs in this PR would be documentation phrasing and grammar, which I find would be too laborious to spell out here individually. I'll do a commit for the PR fixing those issues later, if that's alright.

No problem at all, that's even encouraged.

fiftydinar committed:
chore(default-flatpaks): Make config organization cleaner

Don't use echos for writing configs, use files instead for easier & more intuitive editing.

fiftydinar committed:
fix(default-flatpaks): Copy notification file properly

Made a mistake in variable name

Commit:
docs: README fixes

- don't use ambiguous term "live-user"
- capitalize Flatpak
- grammar and phrasing changes

Commit:
docs: grammar and phrasing changes in configuration file docs

- replace "maintainer's config" with "image's default config"
- rewrite flatpak list format explanation

Commit:
docs: slightly reword local modification section again

fiftydinar committed:
chore(default-flatpaks): Add a missing newline to system flatpak list
